Fix failure in L1Trigger/TrackFindingTracklet reported in ASAN

mrodozov opened the pull request:
#### PR description:

Fixes at least one unit test in ASAN IB 
SLHCUpgradeSimulations/Geometry -> testPhase2PixelNtuple
Using what #30913 suggests

#### PR validation:

the unit test runs

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-31555/18572

- This PR adds an extra 20KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  L1Trigger/TrackFindingTracklet/src/FitTrack.cc  modified in PR(s): #30847, #31554

cmsbuild commented:
A new Pull Request was created by  @mrodozov (Mircho Rodozov) for master.

It involves the following packages:

L1Trigger/TrackFindingTracklet


@cmsbuild, @rekovic, @benkrikler can you please review it and eventually sign? Thanks.
@Martin-Grunewald, @erikbutz this is something you requested to watch as well.
@silviodonato, @dpiparo, @qliphy you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

mmusich commented:
Duplicate of https://github.com/cms-sw/cmssw/pull/31554 ?
In any case the PR title is misleading as the package involved is actually L1Trigger/TrackFindingTracklet and the unit test just happens to consume it.

mrodozov commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
-1

Tested at: efd2a1988ebae802a1472cbdc95e36e1cd6720b5

CMSSW: CMSSW_11_2_X_2020-09-23-1100
SCRAM_ARCH: slc7_amd64_gcc820
You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-bdd106/9526/summary.html

I found follow errors while testing this PR

**Failed tests:  RelVals**

* **RelVals**:

The relvals timed out after 4 hours.

cmsbuild commented:
Comparison not run due to runTheMatrix errors (RelVals and Igprof tests were also skipped)

Commit:
Fix failure in L1Trigger/TrackFindingTracklet

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-31555/18585

- This PR adds an extra 20KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  L1Trigger/TrackFindingTracklet/src/FitTrack.cc  modified in PR(s): #30847, #31554

cmsbuild commented:
Pull request #31555 was updated. @cmsbuild, @rekovic, @benkrikler can you please check and sign again.

mrodozov commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

mrodozov commented:
> Duplicate of #31554 ?
> In any case the PR title is misleading as the package involved is actually L1Trigger/TrackFindingTracklet and the unit test just happens to consume it.


yes it is but I've removed the redundancies and changed the misleading title.

mrodozov commented:
abort
got stuck again

mrodozov commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

smuzaffar commented:
@mrodozov , I think due to this change it is going in to some infinite loop. Please run wf 23234.0 locally to see if it finishes

smuzaffar commented on L1Trigger/TrackFindingTracklet/src/FitTrack.cc:
```diff
@@ -933,48 +933,40 @@ void FitTrack::execute() {
-        nMatches++;
-        match = true;
-      }
+    while (indexArray[2] < matches2.size() && matches2[indexArray[2]] == bestTracklet) {
```
@mrodozov , this is bug. Please use `matches3`

smuzaffar commented on L1Trigger/TrackFindingTracklet/src/FitTrack.cc:
```diff
@@ -933,48 +933,40 @@ void FitTrack::execute() {
-        nMatches++;
-        match = true;
-      }
+    while (indexArray[3] < matches2.size() && matches2[indexArray[3]] == bestTracklet) {
```
same here, use `matches4`

mrodozov commented:
abort

cmsbuild commented:
Jenkins tests are aborted.

Commit:
Update FitTrack.cc

cmsbuild commented:
The code-checks are being triggered in jenkins.

mrodozov commented:
please test

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-31555/18597

- This PR adds an extra 20KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  L1Trigger/TrackFindingTracklet/src/FitTrack.cc  modified in PR(s): #30847, #31554

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
Pull request #31555 was updated. @cmsbuild, @rekovic, @benkrikler can you please check and sign again.

cmsbuild commented:
+1
Tested at: fec21a80dd19a497e94aa71993ffc4dc429f064b
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-bdd106/9545/summary.html
CMSSW: CMSSW_11_2_X_2020-09-23-2300
SCRAM_ARCH: slc7_amd64_gcc820

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-bdd106/9545/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 35
 - DQMHistoTests: Total histograms compared: 2539438
 - DQMHistoTests: Total failures: 1
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2539415
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 34 files compared)
 - Checked 149 log files, 22 edm output root files, 35 DQM output files

smuzaffar commented:
@rekovic, @benkrikler can you please review this?
@qliphy , @silviodonato this fixes most of ASAN IBs relvals failures. This change here does not change the ebhaviour of code. It just makes sure that we check for the size before accessing the memory.

rekovic commented:
+1

rekovic commented:
Attn @skinnari @tomalin

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @silviodonato, @dpiparo, @qliphy (and backports should be raised in the release meeting by the corresponding L2)

silviodonato commented:
+1
